### DISCUSSION TITLE
External table from PostgresSQL - Query returns all types as text

### DISCUSSION THREAD

**Discussion by igjohemiblti** [Budibase/budibase#2676] [Help]
Hi,

I am working with a PostgreSQL database. Because I do not use schema public, I generated a query to access the desired data, but this issue can be reproduced also inside the "default" public schema.

Let's assume the following table: 

CREATE  TABLE "public".test 
( 
    id bigint DEFAULT nextval('test_id_seq'::regclass) NOT NULL ,
    numbers bigint  NOT NULL ,
    texting varchar(250)  NOT NULL ,
    CONSTRAINT pk_test_id PRIMARY KEY ( id )
 );

Accessing the table without queries treats the columns per data types as expected.
However, when using a query, all the columns are recognized as of text type:

SELECT * FROM public.test;

![SELECT as text](https://user-images.githubusercontent.com/25967589/134020168-17d0eb80-bf1d-49aa-8d29-69b381f9f701.png)

Is there a way to cast data types in order to be able to, for example, order properly according to a number column?

Thanks!

**Comment by dncpax** [comment]
You can change the type using the dropdown in front of each field, then save the query.
I'm not sure if this persists, since every time you run the query field types are reset to text.

**Comment by igjohemiblti** [comment]
Hi! 

_"You can change the type using the dropdown in front of each field, then save the query."_
Tested and not working. 

After changing the field to number and saving the query, the ordering for the data provider (DESC for the column of interest) keeps showing 9 as first item, when I have numbers such as 100).

However, once you click on the table header for the given column, the ordering is correct, but the default ordering, which I assume is given by the data provider, remains as text and not as number.

**Comment by mjashanks** [comment]
Hi @igjohemiblti 

It's possible to manage the ordering in a custom query, using parameters.

Query setup
1. Add an `OrderBy` parameter (you can call it what you like)
2. Use this inside your query like `select * from test orderby {{ OrderBy }}`
3. In your design section, add a Form component
4. Then add an options picker, with a "Custom" schema. Then "Define Values" - these will be our Order by options
5. Add a data provider, reading from your query
6. Click the little cog button (beside the data selection) to pass the value from your options picker into your query
7. Add a table. The table will get refreshed everytime you select a value from the picker

https://user-images.githubusercontent.com/3524181/134155591-9b16f71d-b6c4-4f4e-9c7c-8d701789fb40.mp4

**Reply by AlexSapple** [reply]
Hi all, 

I'm trying to do something almost identical to the parameters suggestion above - using an SQL data source but not having much luck. 

I've a bindings set up for a couple of user inputs to populate a query - looks like this;

![image](https://user-images.githubusercontent.com/30351641/164031212-4f6198f5-b96c-46b1-8461-2100495f47cc.png)

when selecting month/year fields - the table does not refresh as it does in the above video and so it looks like either there is no onChange type event happening or something isn't quite hooked up - I'm new to budibase and so I may well have missed something fundamental. Any suggestion would be appreciated - thanks!

**Reply by AlexSapple** [reply]
in fact - if I add a button to do the refresh - it also doesn't seem to behave as I would have expected - If I add execute query as an action, it does appear to grab the form values, but the table isn't updated based on the response; 

![image](https://user-images.githubusercontent.com/30351641/164036064-143fc3fa-1b7e-483e-9d6c-6d4f9ccebc4b.png)
(as per default values)

changed values that gives no data response but the table doesn't update

![image](https://user-images.githubusercontent.com/30351641/164036508-595b72b8-599b-4785-a6ea-eb91d0c5fdcf.png)

![image](https://user-images.githubusercontent.com/30351641/164036281-83b9871a-df8a-413e-9806-c1a2eaac73e5.png)
